nixos/modules/security/pam.nix

```diff
@@ -460,6 +460,94 @@ let
         description = lib.mdDoc "Contents of the PAM service file.";
       };
 
+      types = mkOption {
+        # This option is experimental and subject to breaking changes without notice.
+        visible = false;
+
+        description = lib.mdDoc ''
+          PAM rules for this service.
+
+          Attribute keys are the `type` of the rules (one of `account`, `auth`, `password`, `session`).
+        '';
+        type = types.attrsOf (types.submodule {
+          options.rules = mkOption {
+            description = lib.mdDoc ''
+              PAM rules for this service and type.
+
+              Attribute keys are the name of each rule.
+            '';
+            type = types.attrsOf (types.submodule ({ config, ... }: {
+              options = {
+                enable = mkOption {
+                  type = types.bool;
+                  default = true;
+                  description = lib.mdDoc ''
+                    Whether this rule is added to the PAM service config file.
+                  '';
+                };
+                order = mkOption {
+                  type = types.int;
+                  default = 1000;
+                  description = lib.mdDoc ''
+                    Order of this rule in the service file. Rules are arranged in ascending order of this value.
+
+                    ::: {.warning}
+                    The `order` values for the built-in rules are subject to change. If you assign a constant value to this option, a system update could silently reorder your rule. You could be locked out of your system, or your system could be left wide open. If you must use this option, set it to a relative offset from another rule's `order` value:
+
+                    ```nix
+                    {
+                      security.pam.services.login.types.auth.rules.foo.order =
+                        config.security.pam.services.login.types.auth.rules.unix.order + 1;
```

Comment:
I think we can have an `order`-based implementation for now, but the current one is too unsafe, because it orders rules arbitrarily if they have the same `order` value. To make this safer, an error should be thrown in such cases, therefore requiring definitions to provide one. There also shouldn't be a default order then.

For the future: The problem of such number based orderings is that it requires a sort of centralised registry to decide which modules should use which numbers, which then doesn't work well as soon as you have third-party modules.

This is what a DAG-based ordering solves, and I agree with others that we should really have that in the future. In fact the docs here suggesting the use of other attributes `order` value is almost the same thing. Note that I have [a PR](https://github.com/NixOS/nixpkgs/pull/97392) open to add a generic DAG-based type to Nixpkgs, though I don't have the time to pick it up right now, would appreciate others to do that.

Reply:
Removed the default order and added a check for equal order values in ee5dd1dd40fa77339898580bb614db920b5af46d.